File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if(head == nullptr || left == right) return head;

        ListNode dummy(0);
        dummy.next = head;
        ListNode* prev = &dummy;

        for(int i = 1; i < left; i++) {
            prev = prev->next;
        }

        ListNode* curr = prev->next;
        for(int i = 0; i < right - left; i++) {
            ListNode* nextTmp = curr->next;
            curr->next = nextTmp->next;
            nextTmp->next = prev->next;
            prev->next = nextTmp;
        }

        return dummy.next;
    }
};

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, left int, right int) *ListNode {
    if head == nil || left == right {
        return head
    }

    dummy := &ListNode{Next: head}
    prev := dummy

    for i := 1; i < left; i++ {
        prev = prev.Next
    }

    curr := prev.Next
    for i := 0; i < right - left; i++ {
        nextTmp := curr.Next
        curr.Next = nextTmp.Next
        nextTmp.Next = prev.Next
        prev.Next = nextTmp
    }

    return dummy.Next
}